File: IsBalanced.py
# Time Complexity: O(N)
# Space Complexity: O(H)
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
      def helper(node, height):
        if node is None:
          return True, height
        
        bal_l, h_l = helper(node.left, height + 1)
        bal_r, h_r = helper(node.right, height + 1)
        bal = (h_l - h_r) <= 1 if h_l > h_r else (h_r - h_l) <= 1

        return bal_l and bal_r and bal, max(h_l, h_r)
      
      return helper(root, 0)[0]

File: src/IsBalanced.java
//TC O(n)
// SC O(1)
public class IsBalanced {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }

        public boolean isBalanced(TreeNode root) {
            if (root == null) {
                return true;
            }
            return findHeight(root) != -1;
        }

        private int findHeight(TreeNode root) {
            if (root == null) {
                return 0;
            }
            int left = findHeight(root.left);
            int right = findHeight(root.right);
            if (left == -1 || right == -1 || Math.abs(left - right) > 1) {
                return -1;
            }
            return Math.max(left, right) + 1;
        }
    }
}
